### PR TITLE
[AGFS fee reform] Defendants & Offences page

### PR DESCRIPTION
**Pull request by lostie**
#### What

- Re-style the defendants page according to the updated designs
- Add behaviour to collect the list of offences based on what type of fee scheme the claim is associated with (in regards to the new fee reform one, there's some extra work that needs to be put in to place once the @colinbruce SPIKE is completed)
- Add behaviour to display the appropriate offences page based on what type of claim and fee scheme associated with the claim

#### Ticket

💎  [PT](https://www.pivotaltracker.com/story/show/155368005)

#### Notes

- The `release date` for the AGFS fee reform was set (for now) to be the **1st March 2018** so that these changes can be tested in the staging/demo environments (AGFS claims with an earliest representation date on or after the `release date` should have a different set of offences list and a different page to display them.

### CODE REVIEW

File: config/settings.yml [hunk 1]
```diff
@@ -80,6 +80,9 @@ feature_flags_enabled?: <%= !%w(test api-sandbox gamma).include?(ENV['ENV']) %>
 active_features:
   - :agfs_fee_reform
 
+# TODO: Update date to actual release date
+agfs_fee_reform_release_date: <%= Date.new(2018, 3, 1) %>
```
Comment (colinbruce) [inline config/settings.yml:84]: Could this be abstracted to something like:
```ruby
<%= ENV['SCHEME_TEN_DATE'] || Date.new(2018, 4, 1) %>
```
So we could overwrite on different environments if needed?

**Review by colinbruce — APPROVED** [review]
One minor observation, but otherwise 👍

File: app/assets/javascripts/application.js [hunk 2]
```diff
@@ -91,7 +91,7 @@ if (!String.prototype.supplant) {
 
   $('.defendants').on('cocoon:after-insert', function(e, el){
     var $el = $(el);
-    if($el.hasClass('js-test-defendant')){
+    if($el.hasClass('defendant-details')){
```
Comment (jsugarman) [inline app/assets/javascripts/application.js:94]: 👍 cool, as we discussed in the office if a class is for functional purposes it should not be called `js-test-blah` or anything like it.

File: app/views/external_users/claims/defendants/_representation_order_fields.html.haml [hunk 3]
```diff
@@ -1,24 +1,20 @@
-.nested-fields.indent-fieldset.js-test-rep-order
+.nested-fields.ro-details
```
Comment (jsugarman) [inline app/views/external_users/claims/defendants/_representation_order_fields.html.haml:1]: minor suggestion: i think its clearer (in the css) if the class name is more explicit, such as `rep-order-details`

File: app/views/external_users/claims/defendants/_representation_order_fields.html.haml [hunk 4]
```diff
@@ -1,24 +1,20 @@
...
-    &nbsp;
-  .form-row
 
-    .form-col.form-col-two-thirds.ro-date
-      = f.gov_uk_date_field(:representation_order_date,
-                            legend_text: 'Representation order date',
-                            legend_class: 'govuk-legend',
-                            id: "defendant_#{@defendant_count}_representation_order_#{@representation_order_count}_representation_order_date",
-                            error_messages: gov_uk_date_field_error_messages(@error_presenter, "defendant_#{@defendant_count + 1}_representation_order_#{@representation_order_count}_representation_order_date"))
+  .form-group.ro-date
```
Comment (jsugarman) [inline app/views/external_users/claims/defendants/_representation_order_fields.html.haml:7]: as above

File: features/page_objects/claim_form_page.rb [hunk 5]
```diff
@@ -27,21 +27,21 @@ class ClaimFormPage < SitePrism::Page
...
 
-  sections :defendants, "div.defendants > div.js-test-defendant" do
+  sections :defendants, ".defendant-details" do
     element :first_name, "div.first-name input"
     element :last_name, "div.last-name input"
 
     section :dob, CommonDateSection, 'div.dob'
 
-    sections :representation_orders, "div.js-test-rep-order" do
+    sections :representation_orders, ".ro-details" do
```
Comment (jsugarman) [inline features/page_objects/claim_form_page.rb:36]: as previous suggestion

File: features/page_objects/claim_form_page.rb [hunk 6]
```diff
@@ -27,21 +27,21 @@ class ClaimFormPage < SitePrism::Page
...
       section :date, CommonDateSection, 'div.ro-date'
```
Comment (jsugarman) [inline features/page_objects/claim_form_page.rb:37]: I see this was already in place 😢

File: features/page_objects/claim_form_page.rb [hunk 7]
```diff
@@ -27,21 +27,21 @@ class ClaimFormPage < SitePrism::Page
...
       element :maat_reference, "div.maat input"
     end
 
     element :add_another_representation_order, "div.links > a"
   end
 
-  element :add_another_defendant, "div.defendants > div:nth-of-type(2) > a.add_fields"
+  element :add_another_defendant, ".defendants-actions a.add_fields"
```
Comment (jsugarman) [inline features/page_objects/claim_form_page.rb:44]: 👍

File: spec/models/fee_scheme_spec.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    context 'for a AGFS claim' do
+      let(:claim) { build(:advocate_claim) }
+
+      context 'when the AGFS fee reform feature is not active' do
+        before do
+          allow(FeatureFlag).to receive(:active?).with(:agfs_fee_reform).and_return(false)
+        end
+
+        it 'returns the default scheme' do
+          expect(fee_scheme).to eq('default')
```
Comment (jsugarman) [inline spec/models/fee_scheme_spec.rb:24]: could use `is_expected.to eq('default')` but no biggee

File: spec/models/fee_scheme_spec.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+          let(:representation_order) { instance_double(RepresentationOrder) }
+          let(:release_date) { 3.months.ago.to_date }
+
+          before do
+            expect(Settings).to receive(:agfs_fee_reform_release_date).and_return(release_date.to_s)
+            expect(claim).to receive(:earliest_representation_order).and_return(representation_order)
+            expect(representation_order).to receive(:representation_order_date).and_return(release_date - 1.month)
+          end
+
+          specify { expect(fee_scheme).to eq('default') }
```
Comment (jsugarman) [inline spec/models/fee_scheme_spec.rb:62]: as previous suggestion

File: spec/models/fee_scheme_spec.rb [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+          let(:representation_order) { instance_double(RepresentationOrder) }
+          let(:release_date) { 3.months.ago.to_date }
+
+          before do
+            expect(Settings).to receive(:agfs_fee_reform_release_date).and_return(release_date.to_s)
+            expect(claim).to receive(:earliest_representation_order).and_return(representation_order)
+            expect(representation_order).to receive(:representation_order_date).and_return(release_date)
+          end
+
+          specify { expect(fee_scheme).to eq('fee_reform') }
```
Comment (jsugarman) [inline spec/models/fee_scheme_spec.rb:75]: as previous suggestion

File: spec/models/fee_scheme_spec.rb [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+          let(:representation_order) { instance_double(RepresentationOrder) }
+          let(:release_date) { 3.months.ago.to_date }
+
+          before do
+            expect(Settings).to receive(:agfs_fee_reform_release_date).and_return(release_date.to_s)
+            expect(claim).to receive(:earliest_representation_order).and_return(representation_order)
+            expect(representation_order).to receive(:representation_order_date).and_return(release_date + 2.days)
+          end
+
+          specify { expect(fee_scheme).to eq('fee_reform') }
```
Comment (jsugarman) [inline spec/models/fee_scheme_spec.rb:88]: as previous suggestion

File: spec/services/claims/fetch_eligible_offences_spec.rb [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  subject(:offences) { described_class.for(claim) }
+
+  shared_examples_for 'a claim with default offences' do
+    context 'and the claim has no associated offence' do
+      before do
+        claim.offence = nil
+      end
+
+      it 'returns a list of all available offences' do
+        expect(offences).to match_array(Offence.all)
```
Comment (jsugarman) [inline spec/services/claims/fetch_eligible_offences_spec.rb:13]: ditto

File: spec/services/claims/fetch_eligible_offences_spec.rb [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+    context 'and the claim has an associated offence' do
+      let(:offence) { create(:offence) }
+
+      before do
+        claim.offence = offence
+      end
+
+      it 'returns a list containing only the associated offence' do
+        expect(offences).to match_array([offence])
```
Comment (jsugarman) [inline spec/services/claims/fetch_eligible_offences_spec.rb:25]: ditto

File: spec/services/claims/fetch_eligible_offences_spec.rb [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        include_examples 'a claim with default offences'
+      end
+
+      context 'and fee scheme for the claim is the AGFS reform scheme' do
+        before do
+          allow(claim).to receive(:fee_scheme).and_return('fee_reform')
+        end
+
+        it 'TODO: returns the eligible offences' do
+          expect(offences).to match_array([:todo])
```
Comment (jsugarman) [inline spec/services/claims/fetch_eligible_offences_spec.rb:66]: ditto

**Review by jsugarman — APPROVED** [review]
👍 LGTM, just some suggestions

**Comment by jsugarman** [comment]
feel free to merge if the only failing test is the total-coverage problem

File: config/settings.yml (hunk 1 above)
Comment (lostie) [reply config/settings.yml:84]: Definitely, I didn't go with that approach until we really needed, as it also requires updates on the deploy repo.

**Pull request merged**